Apps/Sandcastle/gallery/Image-Based Lighting.html

```diff
@@ -66,13 +66,10 @@
         }
 
         var environmentMapURL =
-          "https://cesium.com/assets/kiara_6_afternoon_2k_ibl.ktx";
+          "../../SampleData/EnvironmentMap/kiara_6_afternoon_2k_ibl.ktx2";
         var modelURL = "../../SampleData/models/Pawns/Pawns.glb";
 
-        // This environment map was processed using Google's Filament project. To process your own:
-        // 1 - Download the Filament release (https://github.com/google/filament/releases).
-        // 2 - Run `cmgen --type=ktx --deploy=/path/to/output /path/to/image.hdr`. Other formats are also supported. Run `cmgen --help` for all options.
-        // 3 - Take the generated coefficients and the KTX file and load them in CesiumJS as shown below.
+        // This environment map was processed using Google's Filament project.
 
```

Comment:
Are these instructions out of date/not working? 

One problem we had when initially creating this Sandcastle is it was completely opaque what all these coefficients meant, or how to get your own. The user was never supposed to do this process manually. The glTF IBL extension just wasn't ready at the time.

But now it is: https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Vendor/EXT_lights_image_based#ext_lights_image_based. 

This example can be greatly simplified into essentially one line if we just add support for that extension. I would suggest we do that, and/or just remove this example because it isn't very useful to Cesium users as-is.

Reply:
Yeah, Google's Filament hasn't been updated to support KTX2. I had to rebuild this cubemap by using [this website](https://jaxry.github.io/panorama-to-cubemap/), which gave six PNG images, which I then had adjust the opacity of and then convert to KTX2 using the command line tool from [this repo](https://github.com/KhronosGroup/KTX-Software).

We could always open up an issue to build a tool that automates this?

Reply:
One limitation of using [`EXT_lights_image_based`](https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Vendor/EXT_lights_image_based#ext_lights_image_based) is that it needs to be added to the glTF's individually whereas Cesium users might want to set an environment map for the entire scene.

Reply:
Yes.  It's critical for users to be able to select their own HDR, IBL environments.  Can't expect ordinary models to bundle them.

Reply:
Would most users know how to/want to manually specify the 9 coefficients as shown here? That was the part I was thinking we could omit from this sandcastle, and fallback to the procedural.

Reply:
If you fall back to procedural, the diffuse lighting on the model won't match the specular lighting environment, which is not what a user who cares about lighting environments is going to want to see.  That said, I'm no fan of manually specifying spherical harmonics in code, that seems ridiculous and error-prone.  But is the alternative to force all users to always accept a default procedural diffuse lighting environment?

By comparison, BabylonJS stores pre-filtered lighting environments in `.env` files of their own design, which users can create in their Sandbox ([instructions](https://doc.babylonjs.com/how_to/use_hdr_environment)).  The .env file contains both diffuse and specular lighting pre-filtered, making it ready to render models with.  I don't know if the format is standard or open, but I suppose we could ask.  For users this is very easy, just load the env and assign it to the scene, and all your models will have the correct lighting and reflections of the environment.

In ThreeJS, they can now load `*.hdr` files directly.  But these need some pre-filtering at load time, and they have a class called [PMREMGenerator](https://threejs.org/docs/#api/en/extras/PMREMGenerator) that will do exactly this.  This costs some load-time performance, but offers users the convenience of directly using the most common HDR environment format found online.

I like the KTX file, but it looks like the file is only storing the pre-filtered specular lighting environment, and dumping out the diffuse as a bunch of SH coefficients for users to deal with separately.  Unless maybe KTX2 has some way to include the diffuse environment as well?  Ideally there should be a standard environment format that includes both.

Reply:
I've been working on the VSCode glTF preview, trying to get nice IBLs in all the preview windows, and it would be nice to include Cesium in this too.

In [this commit](https://github.com/AnalyticalGraphicsInc/gltf-vscode/commit/9f9aa470ce337ad203e3f38d68c698aeafb4962c) I hacked up a little `sh.txt` file reader, which is the text file that cmgen writes out when it processed an IBL into a KTX.  It loads the Spherical Harmonics and builds the array of Cartesians for them.  Not sure if this will actually get published, but it seems to work.

Reply:
Watching a Virtual SIGGRAPH presentation on Filament (Moving Mobile Graphics, ~2:28:00).  Romain says they're not using spherical harmonics at all anymore.  Instead, they stop the specular IBL mipmaps at 16x16 smallest size with roughness == 1.0 at that size, and they re-use that particular mip level as their diffuse IBL.

I think Cesium should do that too, and ditch the SH coefficients.